chain/rosetta-rpc/src/adapters/transactions.rs

```diff
@@ -1,69 +1,164 @@
 use std::string::ToString;
 
-/// Constructs a transaction identifier from the cause of a state change.
+use near_primitives::hash::CryptoHash;
+
+use actix::Addr;
+
+/// Mapping from NEAR transaction or receipt hash to list of receipts.
+pub(crate) struct ExecutionToReceipts {
+    map: std::collections::HashMap<CryptoHash, Vec<CryptoHash>>,
+}
+
+impl ExecutionToReceipts {
+    /// Fetches execution outcomes for given block and constructs a mapping from
+    /// transaction or receipt causing the execution to list of created
+    /// receipts’ hashes.
+    pub(crate) async fn for_block(
+        view_client_addr: Addr<near_client::ViewClientActor>,
+        block_hash: CryptoHash,
+    ) -> crate::errors::Result<Self> {
+        let map = view_client_addr
+            .send(near_client::GetExecutionOutcomesForBlock { block_hash })
+            .await?
+            .map_err(crate::errors::ErrorKind::InternalInvariantError)?
+            .into_values()
+            .flat_map(|outcomes| outcomes)
+            .filter(|exec| !exec.outcome.receipt_ids.is_empty())
+            .map(|exec| (exec.id, exec.outcome.receipt_ids))
+            .collect();
+        Ok(Self { map })
+    }
+
+    /// Creates an empty mapping.  This is useful for tests.
+    #[cfg(test)]
+    fn empty() -> Self {
+        Self { map: Default::default() }
+    }
+
+    /// Returns list of related transactions for given NEAR transaction or
+    /// receipt.  Note that the hash is removed from the mapping so subsequent
+    /// calls of the method with the same hash will always return an empty
+    /// vector.
```

Comment:
> calls of the method with the same hash will always return an empty vector.

Why is this necessary? I always question myself what was the reason the code does extra work. It seems to me that `get_for_cause` does not rely on the fact that second call should lead to an empty vector.